tensorboard/backend/application.py

```diff
@@ -337,6 +342,56 @@ def _create_wsgi_app(self):
     app = _handling_errors(app)
     return app
 
+  @wrappers.Request.application
+  def _serve_plugin_entry(self, request):
+    """Serves a HTML for iframed plugin entry point.
+
+    Args:
+      request: The werkzeug.Request object.
+
+    Returns:
+      A werkzeug.Response object.
+    """
+    name = request.args.get('name')
+    plugins = [
+        plugin for plugin in self._plugins if plugin.plugin_name == name]
+
+    if not plugins:
+      return http_util.Respond(
+          request, 'Plugin not found', 'text/plain', code=404)
+
+    if len(plugins) > 1:
+      # Technically is not possible as plugin names are unique and is checked
+      # by the check on __init__.
+      raise AssertionError('Plugin invariant error')
+
+    plugin = plugins[0]
+    module_path = plugin.frontend_metadata().es_module_path
+    if not module_path:
+      return http_util.Respond(
+          request, 'Plugin is not module loadable', 'text/plain', code=400)
+
+    # non-self origin is blocked by CSP but this is a good invariant checking.
+    if urlparse.urlparse(module_path).netloc:
+      raise AssertionError('Expected es_module_path to be non-absolute path')
```

Comment:
This should be a `ValueError` (or some other normal error type), not an
`AssertionError`: I can easily construct a plugin that hits this error,
so it can only be a contract violation, not an invariant violation.